libcontainer/dmz/overlayfs_linux.go

```diff
@@ -0,0 +1,133 @@
+// we can create a safe zero-copy sealed version of /proc/self/exe.
+func sealedOverlayfs(binPath, tmpDir string) (_ *os.File, Err error) {
+	// binPath is going to be /proc/self/exe, so do a readlink to get the real
+	// path. overlayfs needs the real underlying directory for this protection
+	// mode to work properly.
+	if realPath, err := os.Readlink(binPath); err == nil {
+		binPath = realPath
+	}
+	binLowerDirPath, binName := filepath.Split(binPath)
+	// Escape any ":"s or "\"s in the path.
+	binLowerDirPath = escapeOverlayLowerDir(binLowerDirPath)
+
+	// Overlayfs requires two lowerdirs in order to run in "lower-only" mode,
+	// where writes are completely blocked. Ideally we would create a dummy
+	// tmpfs for this, but it turns out that overlayfs doesn't allow for
+	// anonymous mountns paths.
+	// NOTE: I'm working on a patch to fix this but it won't be backported.
+	dummyLowerDirPath := escapeOverlayLowerDir(tmpDir)
+
+	overlayCtx, err := fsopen("overlay", unix.FSOPEN_CLOEXEC)
+	if err != nil {
+		return nil, err
+	}
+	defer overlayCtx.Close()
+
+	// Configure the lowerdirs. The binary lowerdir needs to be on the top to
+	// ensure that a file called "runc" (binName) in the dummy lowerdir doesn't
+	// mask the binary.
+	lowerDirStr := binLowerDirPath + ":" + dummyLowerDirPath
+	if err := unix.FsconfigSetString(int(overlayCtx.Fd()), "lowerdir", lowerDirStr); err != nil {
+		return nil, fmt.Errorf("fsconfig set overlayfs lowerdir=%s: %w", lowerDirStr, err)
+	}
+
+	// Get an actual handle to the overlayfs.
+	if err := unix.FsconfigCreate(int(overlayCtx.Fd())); err != nil {
+		return nil, os.NewSyscallError("fsconfig create overlayfs", err)
+	}
+	overlayFd, err := fsmount(overlayCtx, unix.FSMOUNT_CLOEXEC, unix.MS_RDONLY|unix.MS_NODEV|unix.MS_NOSUID)
+	if err != nil {
+		return nil, err
+	}
+	defer overlayFd.Close()
+
+	// Grab a handle to the binary through overlayfs.
+	exeFile, err := utils.Openat(overlayFd, binName, unix.O_PATH|unix.O_NOFOLLOW|unix.O_CLOEXEC, 0)
+	if err != nil {
+		return nil, fmt.Errorf("open %s from overlayfs (lowerdir=%s): %w", binName, lowerDirStr, err)
+	}
+	defer func() {
+		if Err != nil {
+			_ = exeFile.Close()
+		}
+	}()
+
+	// Check that the file is what we expect. Ideally we might check the hash
+	// of the file against /proc/self/exe, but that would require us to copy
+	// the binary (negating the benefit of using overlayfs over memfd-copying).
+	// So instead we just check that the inode number is the same. In theory
+	// this could result in us allowing an incorrect binary in some scenarios
+	// but this would require an attacker to modify /usr/sbin in the host
+	// filesystem -- at which point you've already lost.
+	procSelfStat, err := fstatat(nil, "/proc/self/exe", 0)
+	if err != nil {
+		return nil, err
+	}
+	exeFileStat, err := fstatat(exeFile, "", 0)
+	if err != nil {
+		return nil, err
+	}
+	if procSelfStat.Ino != exeFileStat.Ino {
```

Comment:
But they should be equal, I don't know why there are not equal in my machine.

Reply:
This check might be a bit brittle because it might depend on the lower filesystem type. On btrfs this check works without issue. It might make sense to just remove it, I guess. There isn't a practical attack to worry about anyway...

Reply:
Yes, this should be removed.

Reply:
Okay, so the issue is that overlayfs will re-use inode numbers if all the layer are on the same filesystem. If they are on different filesystems it generates new numbers to avoid collisions. See `ovl_map_dev_ino`.